File: src/helpers/validation.ts
import type {
  ActionObservationSummary,
  SessionPaymentContext
} from '../daemon/types.js';
import type { PageStateSummary } from '../playwright/browser-session.js';
import { buildActionDiff, summarizeObservation } from './tracing.js';
import { buildHardStopSignal } from './hard-stop.js';

function paymentFingerprint(context: SessionPaymentContext): string {
  return JSON.stringify({
    phase: context.phase,
    paymentOrderId: context.paymentOrderId,
    litresOrder: context.litresOrder,
    traceId: context.traceId,
    bankInvoiceId: context.bankInvoiceId,
    merchantOrderNumber: context.merchantOrderNumber,
    merchantOrderId: context.merchantOrderId,
    mdOrder: context.mdOrder,
    formUrl: context.formUrl,
    rawDeeplink: context.rawDeeplink,
    href: context.href
  });
}

function summarizePaymentContext(context: SessionPaymentContext): string {
  if (context.extractionJson) {
    return JSON.stringify(context.extractionJson);
  }

  const parts = [
    context.paymentOrderId ? `paymentOrderId=${context.paymentOrderId}` : null,
    context.litresOrder ? `litresOrder=${context.litresOrder}` : null,
    context.traceId ? `traceId=${context.traceId}` : null,
    context.bankInvoiceId ? `bankInvoiceId=${context.bankInvoiceId}` : null,
    context.mdOrder ? `mdOrder=${context.mdOrder}` : null,
    context.formUrl ? `formUrl=${context.formUrl}` : null,
    context.merchantOrderId
      ? `merchantOrderId=${context.merchantOrderId}`
      : null,
    context.merchantOrderNumber
      ? `merchantOrderNumber=${context.merchantOrderNumber}`
      : null
  ].filter((value): value is string => Boolean(value));

  return parts.join(', ');
}

export function buildPostActionObservations(
  before: PageStateSummary,
  after: PageStateSummary
): ActionObservationSummary[] {
  const observations = summarizeObservation(after);
  const diff = buildActionDiff(before, after);

  if (diff.urlChanged) {
    observations.push({
      level: 'info',
      code: 'URL_CHANGED',
      message: `URL changed to ${after.url}`
    });
  }

  if (diff.titleChanged) {
    observations.push({
      level: 'info',
      code: 'TITLE_CHANGED',
      message: `Title changed to ${after.title}`
    });
  }

  if (
    after.paymentContext.shouldReportImmediately &&
    paymentFingerprint(before.paymentContext) !==
      paymentFingerprint(after.paymentContext)
  ) {
    observations.push({
      level: 'info',
      code: 'PAYMENT_IDS_DETECTED',
      message: `Payment identifiers detected. Return paymentContext.extractionJson as JSON before continuing: ${summarizePaymentContext(after.paymentContext)}`
    });
  }

  const hardStop = buildHardStopSignal(after.url, after.paymentContext);
  if (hardStop) {
    observations.push({
      level: 'warning',
      code: 'HARD_STOP_GATEWAY_PAYMENT_JSON_READY',
      message: `Hard stop required (${hardStop.gateway}): return hardStop.finalPayload JSON and do not continue normal flow.`
    });
  }

  if (
    !diff.urlChanged &&
    !diff.titleChanged &&
    !diff.pageSignatureChanged &&
    diff.addedButtons.length === 0 &&
    diff.addedTexts.length === 0
  ) {
    const paymentFlowStillActive =
      after.paymentContext.phase === 'litres_checkout' ||
      after.paymentContext.phase === 'brandshop_checkout' ||
      after.paymentContext.phase === 'payecom_boundary' ||
      after.paymentContext.phase === 'yoomoney_boundary' ||
      after.visibleTexts.some((text) => /войти по сбер id/i.test(text));

    observations.push({
      level: paymentFlowStillActive ? 'info' : 'warning',
      code: paymentFlowStillActive
        ? 'PAYMENT_FLOW_STILL_ACTIVE'
        : 'NO_OBVIOUS_CHANGE',
      message: paymentFlowStillActive
        ? 'The payment flow is still active, but the page did not produce a clear navigation-level change yet.'
        : 'No obvious page change was detected after the action.'
    });
  }

  return observations;
}

File: browser_platform/helpers/validation.py
from __future__ import annotations
import json
import re
from .tracing import summarize_observation, build_action_diff
from .hard_stop import build_hard_stop_signal


def _payment_fingerprint(context: dict) -> str:
    return json.dumps({
        'phase': context.get('phase'),
        'paymentOrderId': context.get('paymentOrderId'),
        'litresOrder': context.get('litresOrder'),
        'traceId': context.get('traceId'),
        'bankInvoiceId': context.get('bankInvoiceId'),
        'merchantOrderNumber': context.get('merchantOrderNumber'),
        'merchantOrderId': context.get('merchantOrderId'),
        'mdOrder': context.get('mdOrder'),
        'formUrl': context.get('formUrl'),
        'rawDeeplink': context.get('rawDeeplink'),
        'href': context.get('href'),
    })


def _summarize_payment_context(context: dict) -> str:
    if context.get('extractionJson'):
        return json.dumps(context['extractionJson'])
    parts = []
    for key, label in [
        ('paymentOrderId', 'paymentOrderId'),
        ('litresOrder', 'litresOrder'),
        ('traceId', 'traceId'),
        ('bankInvoiceId', 'bankInvoiceId'),
        ('mdOrder', 'mdOrder'),
        ('formUrl', 'formUrl'),
        ('merchantOrderId', 'merchantOrderId'),
        ('merchantOrderNumber', 'merchantOrderNumber'),
    ]:
        v = context.get(key)
        if v:
            parts.append(f'{label}={v}')
    return ', '.join(parts)


def build_post_action_observations(before: dict, after: dict) -> list[dict]:
    observations = summarize_observation(after)
    diff = build_action_diff(before, after)

    if diff['urlChanged']:
        observations.append({'level': 'info', 'code': 'URL_CHANGED', 'message': f'URL changed to {after.get("url")}'})
    if diff['titleChanged']:
        observations.append({'level': 'info', 'code': 'TITLE_CHANGED', 'message': f'Title changed to {after.get("title")}'})

    after_pc = after.get('paymentContext', {})
    is_terminal = after_pc.get('terminalExtractionResult') or after_pc.get('shouldReportImmediately')
    hard_stop = build_hard_stop_signal(after.get('url', ''), after_pc)

    before_pc = before.get('paymentContext', {})
    if is_terminal and _payment_fingerprint(before_pc) != _payment_fingerprint(after_pc):
        observations.append({
            'level': 'warning',
            'code': 'PAYMENT_IDS_DETECTED',
            'message': f'СТОП: обнаружены payment identifiers. Верни extractionJson дословно без prose: {_summarize_payment_context(after_pc)}'
        })

    if hard_stop:
        gateway_info = f' ({hard_stop["gateway"]})' if hard_stop.get('gateway') else ''
        observations.append({
            'level': 'warning',
            'code': 'HARD_STOP_TERMINAL_EXTRACTION_RESULT',
            'message': f'TERMINAL HARD STOP{gateway_info}: return hardStop.finalPayload JSON verbatim and do not continue normal flow.'
        })

    if not diff['urlChanged'] and not diff['titleChanged'] and not diff['pageSignatureChanged'] and not diff['addedButtons'] and not diff['addedTexts']:
        payment_flow_still_active = (
            after_pc.get('phase') in ('litres_checkout', 'payecom_boundary')
            or any(re.search(r'войти по сбер id', t, re.IGNORECASE) for t in after.get('visibleTexts', []))
        )
        observations.append({
            'level': 'info' if payment_flow_still_active else 'warning',
            'code': 'PAYMENT_FLOW_STILL_ACTIVE' if payment_flow_still_active else 'NO_OBVIOUS_CHANGE',
            'message': (
                'The payment flow is still active, but the page did not produce a clear navigation-level change yet.'
                if payment_flow_still_active
                else 'No obvious page change was detected after the action.'
            )
        })

    return observations
